### ISSUE TITLE
Use ewoc?

### ISSUE THREAD

**Issue by alphapapa** [jgkamat/matrix-client-el#49]
@jgkamat I just happened to stumble upon this in the Elisp manual:

> The Ewoc package constructs buffer text that represents a structure of Lisp objects, and updates the text to follow changes in that structure. This is like the “view” component in the “model–view–controller” design
paradigm.  Ewoc means “Emacs’s Widget for Object Collections”.

Theoretically that would be ideal for our room buffers.  It could greatly simplify a lot of our code.  I wonder if there's a reason I've never heard of it, and why I've never seen any other packages using it...

**Comment by jgkamat** [comment]
Woah! I haven't ever heard of this before, but it looks like exactly what we
need to remove the buffer traversing.

I'll try to figure out why no one is using it, likely it's because its just
not popular enough, but I'd rather like to find any issues with it (maybe bugs
in older Emacs versions) before we start using it rather than after.

alphapapa writes:

**Comment by alphapapa** [comment]
Cool.  I'll probably do a little experimenting with it.

I think it would still require some customization, e.g. there's nothing like an "insert-node-by-pred" function that we could use to insert messages by timestamp, so we'd still have to implement that ourselves, which we've already done.  And we already have code to update date headers when the date changes (oh, idea: set a timer to do that at midnight).  So if we had discovered it sooner, it might have saved us some work, but now... I'm not sure.  Still might be worth considering.

**Comment by pvagner** [comment]
I have never used jabber.el however I have just found out it also relies on ewoc. Perhaps that's a good pro argument.

**Comment by alphapapa** [comment]
EWOC is being used in Ement.el, which this project is deprecated in favor of.
